core/src/types/writer.rs

```diff
@@ -90,6 +91,56 @@ impl Writer {
         }
     }
 
+    /// Sink into writer.
+    ///
+    /// sink will read data from given streamer and write them into writer
+    /// directly without extra in-memory buffer.
+    ///
+    /// # Notes
+    ///
+    /// - Sink doesn't support to be used with write concurrently.
+    /// - Sink doesn't support to be used without content length now.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use std::io::Result;
+    ///
+    /// use bytes::Bytes;
+    /// use futures::stream;
+    /// use futures::StreamExt;
+    /// use opendal::Operator;
+    ///
+    /// #[tokio::main]
+    /// async fn sink_example(op: Operator) -> Result<()> {
+    ///     let mut w = op
+    ///         .writer_with("path/to/file")
+    ///         .content_length(2 * 4096)
+    ///         .await?;
+    ///     let stream = stream::iter(vec![vec![0; 4096], vec![1; 4096]]).map(Ok);
+    ///     w.sink(2 * 4096, stream).await?;
+    ///     w.close().await?;
+    ///     Ok(())
+    /// }
+    /// ```
+    pub async fn sink<S, T>(&mut self, size: u64, s: S) -> Result<()>
```

Comment:
Since `sink` is an API called directly by users. I think we should make the arguments, i.e. `size` and `s`, clearer by stating their semantics.